controlplane/pkg/selector/match_selector.go

```diff
@@ -63,7 +63,8 @@ func (m *matchSelector) matchEndpoint(nsLabels map[string]string, ns *registry.N
 		for _, destination := range match.GetRoutes() {
 			// Each NSE should be matched against that destination
 			for _, nse := range networkServiceEndpoints {
-				if isSubset(nse.GetLabels(), destination.GetDestinationSelector()) {
+				if isSubset(nse.GetLabels(), destination.GetDestinationSelector()) &&
+					!isSubset(nsLabels, destination.GetDestinationSelector()) {
```

Comment:
I can't see why the requested Labels should be matched against the destination selector. This might work for the particular case but it looks wrong to me.

Reply:
Question they should not be matched,
Please check test example.

Another idea is to have a list of all already used endpoint ids smoother in context, and do not allow to choose same NSE few times.

Reply:
So we want to avoid any kind of "loops" in the graph we build. The problem is that there is no "centralized" graph building but instead each NSMgr does its own selection when the particular Client requests a service. So actually the central "context" is the Endpoint registry in Kubernetes. The only feasible way to avoid loops seems to be to mark the Endpoints in the registry with a "NS Instance ID", but again that will be very hard to coordinate between all the NSMgrs.

Reply:
But if we want to fix the immediate problem, we can at least try to ensure that a Client never gets an Endpoint declared in the same pod.

Reply:
Hmps, At this point we already know we are not first endpoint in chain, so potentially we could remove all initial selections? So firewall will be filtered, Thoughts?